dialog_lib/agents/abstract.py

```diff
@@ -103,6 +104,74 @@ def messages(self):
         return self.memory.messages
 
 
+class AbstractLcelClass(AbstractLLM):
+
+
+    @property
+    def chat_model(self):
+        """
+        builds and returns the chat model for the LCEL
+        """
+        raise NotImplementedError("Chat model must be implemented")
+
+    @property
+    def context_dict(self):
+        """
+        builds and returns the context dictionary for the LCEL
+        """
+        raise NotImplementedError("Context dictionary must be implemented")
+
+    @property
+    def chain(self):
+        """
+        builds and returns the chain for the LCEL
+        """
+        return (
+            self.context_dict | self.prompt | self.chat_model
```

Comment:
do you have some example of usage of this class? From here it looks like `context_dict` refers to a dictionary with some common keys used for RAG prompts, like `question` and `context`. If that is the case, as an abstraction of RAG chains, we could mount this chain referring to the retriever and llm (to repeat very common names in langchain tutorials/examples) like 
```
    return (
        {
            "context": itemgetter("question") | self.retriever | self.combine_documents,
            "question": itemgetter("question"),
        }
        | self.prompt
        | self.llm
        | StrOutputParser()
    )
```
Some notes:
- `self.llm` would be the `self.chat_model`, I'm just showing a common name in several examples, both are good but not all LLM models are indeed _chat_ models
- `self.retriever`: this naming is very important, since it's a very important part of the RAG architecture. Instead of leaving the user to implement it in `context_dict`, it makes clear that a retriever must be implemented to be called here
- `self.combine_documents`: this allows users to implement any function they want to combine documents returned by the retriever. A common example would be 

```
from langchain.schema import format_document
from langchain.prompts.prompt import PromptTemplate
from dialog.settings import vectordb_settings

DEFAULT_DOCUMENT_PROMPT = PromptTemplate.from_template(template="{page_content}")


def combine_documents(
    docs, document_prompt=DEFAULT_DOCUMENT_PROMPT, document_separator="\n\n"
):
    """Combine documents into a single string."""
    doc_strings = [format_document(doc, document_prompt) for doc in docs]
    return document_separator.join(doc_strings)
    ```